### ISSUE TITLE
feat: abstraction to allow customized EVMs

### ISSUE THREAD

**Issue by zobront** [anton-rs/kona#839]
Currently `kona-executor` exposes the revm handler in order to customize block execution. However, larger customizations of the EVM are not possible.

This abstraction already exists in reth/revm, where:
- reth implements `BasicBlockExecutor`, which contains a `strategy`.
- in order to `execute()` ([link](https://github.com/paradigmxyz/reth/blob/36db1c24077f7a500048c545c4c5cefb6a27224a/crates/evm/src/execute.rs#L344-L355)), it calls `apply_pre_execution_changes()`, `execute_transactions()`, `apply_post_execution_changes(), and `finish()` on the strategy.
- reth includes the [`OpExecutionStrategy`](https://github.com/paradigmxyz/reth/blob/main/crates/optimism/evm/src/execute.rs) which is a good example of this implementation.
- All strategies contain an `evm_config` which must implement [`ConfigureEvm`](https://github.com/paradigmxyz/reth/blob/36db1c24077f7a500048c545c4c5cefb6a27224a/crates/evm/src/lib.rs#L44).

This `evm_config` is what allows customization, as it defines three functions that replace some of the hardcoded actions in Kona:
- `fill_cfg_env()` replaces `self.evm_cfg_env()`
- `fill_block_env()` replaces `Self::prepare_block_env()`
- `fill_tx_env()` replaces `Self::prepare_tx_env()`

I believe the best option to accomplish this is to refactor Executor to rely on a struct implementing the `ConfigureEvm` trait in order to perform the above actions. For "normal" Kona, we can break the existing functions into a new type that implements the `ConfigureEvm` trait. For other customized chains like Odyssey, we can simply inherit whatever `ConfigureEvm` struct they are passing to reth for execution.

I want to talk through this and make sure it's the right strategy before touching any code, so let's discuss. Also available for a live call on Monday if that's easier.

**Comment by clabby** [comment]
> I believe the best option to allow Kona more flexibility in execution is to refactor Executor to rely on a struct implementing the ConfigureEvm trait in order to perform the above actions. For "normal" Kona, we can break the existing functions into a new type that implements the ConfigureEvm trait. For other customized chains like Odyssey, we can simply inherit whatever ConfigureEvm struct they are passing to reth for execution.

Yeah, this sounds like a nice minimal diff approach. Though if we have the time / are willing to make the effort, I think it would be great if we could upstream the traits that reth / revm are using for this configurability to `alloy` and/or `op-alloy` so that we can all operate on a common set of abstractions. In the medium-longer term, that buys us quite a bit - modifications to the OP Stack would only have to make their execution layer changes in one place and get `kona` + `reth` compatibility in one go.

> @clabby You had mentioned implementing the traits in alloy / op-alloy. I assume this accomplishes what you wanted, or was there a reason you prefer it to exist there?

Yeah I'd prefer if it wasn't in `kona` or `reth`. A common set of executor abstractions is pretty desirable.

---

All that said, let me know your appetite for making that sort of change. I'm open to taking a more minimal approach at first, but would def prefer if we could work w/ @mattsse et. co on standardizing the interface.

**Comment by clabby** [comment]
Chatted about this a bit more with @mattsse today. Looking like the effort on creating a unified interface has picked back up, issue should be able to be revived soon.

**Comment by clabby** [comment]
Adding https://github.com/op-rs/kona/labels/M-blocked, we're waiting on `ithacaxyz/evm` to be open sourced. cc @mattsse 

Some more info on why this is important to us:
1. Having a shared executor interface between `kona` and `op-reth` will lower our maintenance burden for adding features to the EL a ton.
2. Lowers implementation disparity risk, deduplicating the EL STF code between `op-reth` and `kona`
3. Allows for easy drop-ins of EL STF diffs, such as Odyssey's, to `kona`.

**Comment by clabby** [comment]
> latest here is that alloy-evm abstraction will be used right?

Yeah correct 👍

**Comment by emhane** [comment]
> > latest here is that alloy-evm abstraction will be used right?
> 
> Yeah correct 👍

does that mean the issue is unblocked now?

**Comment by clabby** [comment]
> > > latest here is that alloy-evm abstraction will be used right?
> > 
> > 
> > Yeah correct 👍
> 
> does that mean the issue is unblocked now?

Yep! I'll remove the label.

**Comment by clabby** [comment]
Closing for now with #1400. There's still a bit of work to be done, but `alloy-evm` is a great step in this direction. Let's follow up with more specific asks, if any consumers need more 👍
